src/lm_speed.cpp

```diff
@@ -0,0 +1,502 @@
+// https://cran.r-project.org/web/packages/RcppEigen/vignettes/RcppEigen-Introduction.pdf
+// [[Rcpp::export]]
+Eigen::MatrixXd AtA(const Eigen::MatrixXd& A) {
+  int n(A.cols());
+  return Eigen::MatrixXd(n,n).setZero().selfadjointView<Eigen::Lower>()
+                             .rankUpdate(A.adjoint());
+}
+
+//' @export
+// [[Rcpp::export]]
+Eigen::MatrixXd mult_diag(const Eigen::MatrixXd& x, const Eigen::ArrayXd& d) {
+
+  Eigen::MatrixXd out(x.rows(), x.cols());
+  for (unsigned j = 0; j < x.cols(); ++j) {
+    out.col(j) = x.col(j) * d(j);
+  }
+
+  return out;
+}
+
+//' @export
+// [[Rcpp::export]]
+Eigen::VectorXd extract_vec(const Eigen::VectorXd& full, const Eigen::VectorXi& ind)
+{
+  int num_indices = ind.sum();
+  Eigen::VectorXd target(num_indices);
+  int j = 0;
+  for (int i = 0; i < full.size(); i++)
+  {
+    if (ind(i) == 1) {
+      target(j) = full(i);
+      j++;
+    }
+  }
+  return target;
+}
+
+//' @export
+// [[Rcpp::export]]
+Eigen::MatrixXd extract_mat_rows(const Eigen::MatrixXd& full, const Eigen::VectorXi& ind)
+{
+  int num_indices = ind.sum();
+  Eigen::MatrixXd target(num_indices, full.cols());
+  int j = 0;
+  for (int i = 0; i < full.rows(); i++)
+  {
+    if (ind(i) == 1) {
+      target.row(j) = full.row(i);
+      j++;
+    }
+  }
+  return target;
+}
+
+// extract_mat(const Eigen::DenseBase<Eigen::Matrix>& full, const Eigen::DenseBase<T>& ind)
+// {
+//   int num_indices = ind.innerSize();
+//   target_t target(num_indices);
+//   for (int i = 0; i < num_indices; i++)
+//   {
+//     target[i] = full[ind[i]];
+//   }
+//   return target;
+// }
+
+//' @export
+// [[Rcpp::export]]
+List lm_ei_test(Eigen::Map<Eigen::MatrixXd>& Xfull,
+                const Eigen::Map<Eigen::VectorXd>& y,
+                const Rcpp::Nullable<Rcpp::NumericMatrix> & Xunweighted,
+                const Rcpp::Nullable<Rcpp::NumericVector> & weight,
+                const double & weight_mean,
+                const Rcpp::Nullable<Rcpp::IntegerVector> & cluster,
+                const unsigned & J,
+                const bool & ci,
+                const String type,
+                const std::vector<bool> & which_covs,
+                const bool& chol,
+                const bool& trychol) {
+
+  const int n(Xfull.rows()), p(Xfull.cols());
+  int r = p;
+  Eigen::MatrixXd Xfullt(Xfull.transpose());
+  Eigen::MatrixXd XtX_inv, R_inv, Vcov_hat;
+  Eigen::VectorXd beta_out(Eigen::VectorXd::Constant(p, ::NA_REAL));
+
+  bool full_rank = true;
+
+  // Many of the lines below are inspired by the fastLm function from RcppEigen
+  // https://cran.r-project.org/web/packages/RcppEigen/vignettes/RcppEigen-Introduction.pdf
+  if (chol) {
+    const Eigen::LLT<Eigen::MatrixXd> llt(Xfullt*Xfull);
+    beta_out = llt.solve(Xfull.adjoint() * y);
+    R_inv = llt.matrixL().solve(Eigen::MatrixXd::Identity(p, p));
+    XtX_inv = R_inv.transpose() * R_inv;
+  } else {
```

Comment:
This block is basically just "only run and try cholesky" and I don't think we will need this going forward.